app/lib/metrics_service.rb

```diff
@@ -90,28 +93,23 @@ def self.merge_identities_in_mixpanel(user, cookies={})
   # @param {String} name Name of the event
   # @param {Hash} props Properties associated with the event
   # @param {User} user User model object
-  def self.log(name, props={}, user)
+  def self.log(name, props={}, user=nil)
     Rails.logger.info "#{Time.zone.now}: Logging analytics to Mixpanel for event name: #{name}"
 
     props.merge!({
       appId: 'single-cell-portal',
       env: Rails.env,
-      registeredForTerra: user.registered_for_firecloud,
       logger: 'app-backend'
     })
 
     headers = get_default_headers(user)
 
-    access_token = user.token_for_api_call
-    user_id = user.get_metrics_uuid
-
-    if access_token.nil? || !user.registered_for_firecloud
-      # User is unauthenticated / unregistered / anonymous
-      props['distinct_id'] = user_id
-      headers.delete('Authorization')
-      props['authenticated'] = false
+    # configure properties/headers depending on user presence
+    # only pass user token if user is registered for Terra to avoid 4xx/5xx errors
+    if user.present? && user.registered_for_firecloud
```

Comment:
Blocker: I'm still seeing an error that is preventing reporting to Bard - if I'm not signed in, `user.get_metrics_uuid` throws a `NoMethodError` because there's no user present.

In thinking about this, there's definitely more context that was missing from the ticket.  As I understand it, the way Bard functions in the background is that it talks to SAM to set user identities, and it does so one of two ways: via the `Authorization: Bearer` token, or the `distinct_id` UUID param.  The former is only present if the user is signed in, but the latter is always accessible as a cookie in the request.  If we don't pass either value, Bard will respond `400` as it cannot determine the identity of the user.  Previously, `MetricsService.log` was only ever called in places where we knew we had an authenticated user, but the `too-many-genes` event breaks that paradigm.

If we look at `report_error`, it includes the request object as a parameter so that it can access request cookies (among other things).  I think we need to take this approach to ensure we can always report events - that will ensure we always have a useable credential.  Also - if we refactor the method signature like this:

```
def self.log(name, props={}, user=nil, request: nil)
```

Then the only place we need to update to pass in the request is this case - the `too-many-genes` event.  Everywhere else `MetricsService.log` is called, there is a user present.  Then we can rewrite this block as:

```
if user.present? && user.registered_for_firecloud
  props.merge!({ authenticated: true, registeredForTerra: user.registered_for_firecloud })
else
  distinct_id = user&.get_metrics_uuid || request&.cookies['user_id']
  props.merge!({ authenticated: false, distinct_id: distinct_id, registeredForTerra: false })
end
```

Reply:
Shoot I shouldn't have missed that, thanks for pointing this out, fixed per your suggestion: [89a0295](https://github.com/broadinstitute/single_cell_portal_core/pull/1662/commits/89a0295cc6dbc5c3b592372e872da78c353ca584)

Reply:
Ah I also neglected to mention that the [`search-too-many-genes`](https://github.com/broadinstitute/single_cell_portal_core/blob/development/app/controllers/api/v1/visualization/expression_controller.rb#L172) event also needs to have the request passed in:

```
MetricsService.log('search-too-many-genes', { numGenes: num_genes }, current_api_user, request:)
```

Reply:
geeze nope you totally did, I read too quickly sorry thanks for catching that.